From 9b5cb5cc7214233fb619be915a233e111a05708f Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 09 May 2022 17:53:45 +0300
Subject: [PATCH] [SW-0] rom: reuse lbw and hbw apis

In order to configure lbw and hbw range registers in recovery flow,
apis have been added. These apis have been re-used wherever hardcoded
range register configuration was done earlier.

This would unify the range register configuration implementation in
rom.

Change-Id: I1a45e3dcb154f82a8cdc062f8a661cf3f91ef3fb
---

diff --git a/arch/gaudi3/arc/dev_security.h b/arch/gaudi3/arc/dev_security.h
index 3789e2a..41d8043 100644
--- a/arch/gaudi3/arc/dev_security.h
+++ b/arch/gaudi3/arc/dev_security.h
@@ -83,6 +83,17 @@
 	HL_SEC_HBW_RR_SHORT_7 = 7,
 };
 
+/*
+ * Enum to set the range registers to a predefined constant value.
+ * eg: Open or block the complete lbw or hbw ranges, in this case
+ * the min and max config values are constant (max possible).
+ */
+enum rom_rr_ctx_static_config {
+	RR_OPEN_ALL,
+	RR_BLOCK_ALL,
+	RR_MAX_CTX_TYPES,
+};
+
 int secureboot_verify_image_v2(const uint8_t *image, uint32_t image_len, uint8_t *key,
 				uint32_t key_len, uint8_t *sig, uint32_t sig_len,
 				uint8_t sec_ver, uint8_t key_index);
diff --git a/arch/gaudi3/arc/security.c b/arch/gaudi3/arc/security.c
index 8dc1248..fde240e 100644
--- a/arch/gaudi3/arc/security.c
+++ b/arch/gaudi3/arc/security.c
@@ -94,6 +94,36 @@
 	}
 };
 
+static struct rom_rr_ctx rr_ctx_lbw[RR_MAX_CTX_TYPES] = {
+	[RR_OPEN_ALL] = {
+		.data_min = RR_DEFAULT_VAL,
+		.data_max = RR_DEFAULT_VAL,
+		.en_val = RR_PRIV_RANGE_DIS,
+	},
+	[RR_BLOCK_ALL] = {
+		.data_min = RR_LBW_PRIV_RANGE_MIN_SHORT_VAL,
+		.data_max = RR_LBW_PRIV_RANGE_MAX_SHORT_VAL,
+		.en_val = RR_PRIV_RANGE_EN,
+	},
+};
+
+static struct rom_rr_ctx rr_ctx_hbw[RR_MAX_CTX_TYPES] = {
+	[RR_OPEN_ALL] = {
+		.data_min = RR_DEFAULT_VAL,
+		.data_min_hi = RR_DEFAULT_VAL,
+		.data_max = RR_DEFAULT_VAL,
+		.data_max_hi = RR_DEFAULT_VAL,
+		.en_val = RR_PRIV_RANGE_DIS,
+	},
+	[RR_BLOCK_ALL] = {
+		.data_min = RR_HBW_PRIV_RANGE_MIN_LO_VAL,
+		.data_min_hi = RR_HBW_PRIV_RANGE_MIN_HI_VAL,
+		.data_max = RR_HBW_PRIV_RANGE_MAX_LO_VAL,
+		.data_max_hi = RR_HBW_PRIV_RANGE_MAX_HI_VAL,
+		.en_val = RR_PRIV_RANGE_EN,
+	},
+};
+
 static uint32_t security_rr_calc_rr_dist(struct rom_rr_ctx *rr_ctx)
 {
 	uint32_t rr_dist;
@@ -451,23 +481,6 @@
 void security_smi_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 void security_jtag_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 
-static void security_disable_i2c_s_if(void)
-{
-	/* Disable LBW */
-	regmap_writel(RR_LBW_PRIV_RANGE_MIN_SHORT_VAL,
-			mmD0_I2C_S_MSTR_IF_RR_LBW_PRIV_RANGE_MIN_SHORT_0);
-	regmap_writel(RR_LBW_PRIV_RANGE_MAX_SHORT_VAL,
-			mmD0_I2C_S_MSTR_IF_RR_LBW_PRIV_RANGE_MAX_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_I2C_S_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
-
-	/* Disable HBW */
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_LO_VAL, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_HI_VAL, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_LO_VAL, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_HI_VAL, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_HI_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
-}
-
 static void security_i2c_axprot_config(void)
 {
 	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
@@ -478,37 +491,6 @@
 	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
-}
-
-static void security_enable_i2c_s_if(void)
-{
-	/*
-	 * RR0 is configured by FW and RR7/13 is configured by HW.
-	 * Disable both RR's.
-	 */
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_13);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_EN_7);
-
-	security_i2c_axprot_config();
-}
-
-static void security_disable_smi_if(void)
-{
-	/* Disable LBW */
-	regmap_writel(RR_LBW_PRIV_RANGE_MIN_SHORT_VAL,
-			mmD0_SMI_MSTR_IF_RR_LBW_PRIV_RANGE_MIN_SHORT_0);
-	regmap_writel(RR_LBW_PRIV_RANGE_MAX_SHORT_VAL,
-			mmD0_SMI_MSTR_IF_RR_LBW_PRIV_RANGE_MAX_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_SMI_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
-
-	/* Disable HBW */
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_LO_VAL, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_HI_VAL, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_LO_VAL, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_HI_VAL, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_HI_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 }
 
 static void security_smi_axprot_config(void)
@@ -523,24 +505,22 @@
 	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
 }
 
-static void security_enable_smi_if(void)
-{
-	/*
-	 * RR0 is configured by FW and RR7/13 is configured by HW.
-	 * Disable both RR's.
-	 */
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_13);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_EN_7);
-
-	security_smi_axprot_config();
-}
-
 void security_enable_twi_if(void)
 {
-	security_enable_i2c_s_if();
-	security_enable_smi_if();
+	security_i2c_axprot_config();
+	security_smi_axprot_config();
+
+	rr_ctx_lbw[RR_OPEN_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_0;
+	security_rr_twi_lbw_config(&rr_ctx_lbw[RR_OPEN_ALL]);
+
+	rr_ctx_lbw[RR_OPEN_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_13;
+	security_rr_twi_lbw_config(&rr_ctx_lbw[RR_OPEN_ALL]);
+
+	rr_ctx_hbw[RR_OPEN_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_0;
+	security_rr_twi_lbw_config(&rr_ctx_hbw[RR_OPEN_ALL]);
+
+	rr_ctx_hbw[RR_OPEN_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_7;
+	security_rr_twi_lbw_config(&rr_ctx_hbw[RR_OPEN_ALL]);
 }
 
 void security_disable_twi_if(void)
@@ -548,8 +528,17 @@
 	if ((arcboot_rom_global.config_fuse & EFUSE_SYSCFG_RR_MASK) == 0)
 		return;
 
-	security_disable_i2c_s_if();
-	security_disable_smi_if();
+	rr_ctx_lbw[RR_BLOCK_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_0;
+	security_rr_twi_lbw_config(&rr_ctx_lbw[RR_BLOCK_ALL]);
+
+	rr_ctx_lbw[RR_BLOCK_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_13;
+	security_rr_twi_lbw_config(&rr_ctx_lbw[RR_BLOCK_ALL]);
+
+	rr_ctx_hbw[RR_BLOCK_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_0;
+	security_rr_twi_lbw_config(&rr_ctx_hbw[RR_BLOCK_ALL]);
+
+	rr_ctx_hbw[RR_BLOCK_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_7;
+	security_rr_twi_lbw_config(&rr_ctx_hbw[RR_BLOCK_ALL]);
 }
 
 void security_handle_twi(void)
@@ -573,19 +562,17 @@
 	if ((arcboot_rom_global.config_fuse & EFUSE_SYSCFG_RR_MASK) == 0)
 		return;
 
-	/* Disable LBW */
-	regmap_writel(RR_LBW_PRIV_RANGE_MIN_SHORT_VAL,
-			mmD0_JT_MSTR_IF_RR_LBW_PRIV_RANGE_MIN_SHORT_0);
-	regmap_writel(RR_LBW_PRIV_RANGE_MAX_SHORT_VAL,
-			mmD0_JT_MSTR_IF_RR_LBW_PRIV_RANGE_MAX_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_JT_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
+	rr_ctx_lbw[RR_BLOCK_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_0;
+	security_rr_jtag_lbw_config(&rr_ctx_lbw[RR_BLOCK_ALL]);
 
-	/* Disable HBW */
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_LO_VAL, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_HI_VAL, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_LO_VAL, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_HI_VAL, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_HI_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
+	rr_ctx_lbw[RR_BLOCK_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_13;
+	security_rr_jtag_lbw_config(&rr_ctx_lbw[RR_BLOCK_ALL]);
+
+	rr_ctx_hbw[RR_BLOCK_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_0;
+	security_rr_jtag_hbw_config(&rr_ctx_hbw[RR_BLOCK_ALL]);
+
+	rr_ctx_hbw[RR_BLOCK_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_7;
+	security_rr_jtag_hbw_config(&rr_ctx_hbw[RR_BLOCK_ALL]);
 }
 
 static void security_jtag_axprot_config(void)
@@ -604,14 +591,17 @@
 {
 	security_jtag_dap(SECURITY_JTAG_DAP_ENABLE);
 
-	/*
-	 * RR0 is configured by FW and RR7/13 is configured by HW.
-	 * Disable both RR's.
-	 */
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_13);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
-	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_EN_7);
+	rr_ctx_lbw[RR_OPEN_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_0;
+	security_rr_jtag_lbw_config(&rr_ctx_lbw[RR_OPEN_ALL]);
+
+	rr_ctx_lbw[RR_OPEN_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_13;
+	security_rr_jtag_lbw_config(&rr_ctx_lbw[RR_OPEN_ALL]);
+
+	rr_ctx_hbw[RR_OPEN_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_0;
+	security_rr_jtag_hbw_config(&rr_ctx_hbw[RR_OPEN_ALL]);
+
+	rr_ctx_hbw[RR_OPEN_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_7;
+	security_rr_jtag_hbw_config(&rr_ctx_hbw[RR_OPEN_ALL]);
 
 	security_jtag_axprot_config();
 }
@@ -704,32 +694,14 @@
 
 static void security_rr_lbw_config(void)
 {
-	regmap_writel(RR_LBW_PRIV_RANGE_MIN_SHORT_VAL,
-			mmD0_PCIE_ELBI_RR_MSTR_IF_RR_LBW_PRIV_RANGE_MIN_SHORT_0);
-	regmap_writel(RR_LBW_PRIV_RANGE_MAX_SHORT_VAL,
-			mmD0_PCIE_ELBI_RR_MSTR_IF_RR_LBW_PRIV_RANGE_MAX_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_EN,
-			mmD0_PCIE_ELBI_RR_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
-
-	regmap_writel(RR_LBW_PRIV_RANGE_MIN_SHORT_VAL,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_LBW_PRIV_RANGE_MIN_SHORT_0);
-	regmap_writel(RR_LBW_PRIV_RANGE_MAX_SHORT_VAL,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_LBW_PRIV_RANGE_MAX_SHORT_0);
-	regmap_writel(RR_PRIV_RANGE_EN,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_0);
+	rr_ctx_lbw[RR_BLOCK_ALL].rr_idx = HL_SEC_LBW_RR_SHORT_0;
+	security_rr_pcie_lbw_config(&rr_ctx_lbw[RR_BLOCK_ALL]);
 }
 
 static void security_rr_hbw_config(void)
 {
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_LO_VAL,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MIN_HI_VAL,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_HBW_PRIV_RANGE_MIN_HI_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_LO_VAL,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_LO_0);
-	regmap_writel(RR_HBW_PRIV_RANGE_MAX_HI_VAL,
-			mmD0_PCIE_MSTR_RR_MSTR_IF_RR_HBW_PRIV_RANGE_MAX_HI_0);
-	regmap_writel(RR_PRIV_RANGE_EN, mmD0_PCIE_MSTR_RR_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
+	rr_ctx_hbw[RR_BLOCK_ALL].rr_idx = HL_SEC_HBW_RR_SHORT_0;
+	security_rr_pcie_hbw_config(&rr_ctx_hbw[RR_BLOCK_ALL]);
 }
 
 static void security_rr_config(void)
